Source/IGListDiff.mm

```diff
@@ -228,29 +237,64 @@ static id IGListDiffing(BOOL returnIndexPaths,
+                const NSInteger insertOffset = insertOffsets[i];
+                const NSInteger deleteOffset = deleteOffsets[oldIndex];
+                if ((oldIndex - deleteOffset + insertOffset) != i) {
+                    addMoveFromIndexToIndex(oldIndex, i);
+                }
+            }
+        }
+
+        addIndexToMap(toSection, i, newArray, newMap);
+    }
 
+    if ((option & IGListDiffMovesIncrementally) != 0) {
+        for (NSInteger i = 0; i < newCount; i++) {
+            const IGListRecord record = newResultsArray[i];
+            const NSInteger oldIndex = record.index;
             // calculate the offset and determine if there was a move
             // if the indexes match, ignore the index
-            const NSInteger insertOffset = insertOffsets[i];
-            const NSInteger deleteOffset = deleteOffsets[oldIndex];
-            if ((oldIndex - deleteOffset + insertOffset) != i) {
-                id move;
-                if (returnIndexPaths) {
-                    NSIndexPath *from = [NSIndexPath indexPathForItem:oldIndex inSection:fromSection];
-                    NSIndexPath *to = [NSIndexPath indexPathForItem:i inSection:toSection];
-                    move = [[IGListMoveIndexPath alloc] initWithFrom:from to:to];
-                } else {
-                    move = [[IGListMoveIndex alloc] initWithFrom:oldIndex to:i];
+            const NSInteger insertOffset = oldIndex == NSNotFound ? 0 : oldInsertOffsets[oldIndex];
+            const NSInteger deleteOffset = oldIndex == NSNotFound ? 0 : deleteOffsets[oldIndex];
+            // initial index is array state after applying deletes and insertions
+            const NSInteger initialIndex = oldIndex == NSNotFound ? i : oldIndex - deleteOffset + insertOffset;
+            const NSInteger currentIndex = initialIndex + moveOffsets[initialIndex];
+            const NSInteger deltaToFinalIndex = i - currentIndex;
+            // process moves only if delta < 0
+            // this is simpler, but results in less optimal moves
+            //  0  1  2  3 old
+            // +3 -1 -1 -1 moveOffsets
+            //  1  2  3  0 new
+            // is currently reported as 3 moves
+            // ideally it would be reported as one move
+            if (deltaToFinalIndex < 0) {
+                addMoveFromIndexToIndex(currentIndex, i);
+                // array state at the current step, key is current index, value initial index
+                vector<NSInteger> currentIndexes(newCount);
+                for (NSInteger k = 0; k < newCount; k++) {
+                    currentIndexes[k + moveOffsets[k]] = k;
+                }
+                // update move offsets for moves
+                //  0  1  2 old
+                // +1 +1 -2 moveOffsets
+                //  2  0  1 new
+                // +2 ±0 -2 moveOffsets
+                //  2  1  0 final
+                for (NSInteger j = currentIndex + deltaToFinalIndex; j < currentIndex; j++) {
```

Comment:
Thanks! To begin with, I’ll set up the current algorithm, and will look into the LCS as time allows.